paddle/fluid/imperative/amp_auto_cast.cc

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/imperative/amp_auto_cast.h"
+
+#include <algorithm>
+#include <memory>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+
+#include "paddle/fluid/imperative/layer.h"
+#include "paddle/fluid/imperative/tracer.h"
+#include "paddle/fluid/imperative/variable_wrapper.h"
+
+namespace paddle {
+namespace imperative {
+
+// TODO(zhiqiu): consider to cache casted params for better inference
+// performance.
+
+// The set of ops that support fp16 calculation and are considered numerically -
+// safe and performance critical. These ops are always converted to fp16.
+static std::unordered_set<std::string> g_white_ops;
```

Comment:
global variable in c++ end?

Reply:
Yes, the contents are passed from python end to c++ end when initializing `amp_guard`.

Reply:
Thanks, I use a singleton object to replace these global variable.